### PR TITLE
Check for broken links with remark-lint

### LINKED ISSUE #366: Feature request: Enforce consistent formatting with remark-lint
I'm making a suggestion to add [`remark-lint`](https://github.com/wooorm/remark-lint) to this repo. 

It should help ensure a consistent formatting across all your `*.md` files as well as helping with broken external links or mistyped internal ones (via [remark-lint-no-dead-urls](https://github.com/davidtheclark/remark-lint-no-dead-urls)). 

We could also run it on every PR through Travis, it's very easy to set that up.

Full list of supported rules [here](https://github.com/wooorm/remark-lint/blob/master/doc/rules.md). Let me know if I should open a PR.

### PR DESCRIPTION
**Pull request by RadValentin**
Fixes #366

- [ ] I am an employee of the company mentioned and confirm all included details are correct
- [x] This PR contains housekeeping only (URL edits, copy changes etc)
- [ ] You know your alphabet - company is listed in alphabetical order in the README
- [ ] The company directly hires employees. No bootcamps / freelance sites / etc
- [ ] The company added hires remote employees, or positions are available to remote workers and are clearly illustrated as such
- [ ] A [company profile](https://github.com/remoteintech/remote-jobs/blob/master/company-profiles/example.md) is included - __Required__ for new additions. (This can be a basic outline but at least something please)
- [ ] Any missing parts from the profile, please use `<Unknown>` to allow for future completion
- [ ] __Remote status__ has details regarding how the culture includes remote employees, how the company integrated remote workers, etc
- [ ] __Region__ details any restrictions to applicants based on geography
- [ ] __How to apply__ details the best approach for new applications, page on site where open position are listed, and any other help available for job hunters

### CODE REVIEW

**Comment by RadValentin** [comment]
FYI I'm still working on this, I'll ask for a review when it's ready.

**Comment by RadValentin** [comment]
Note: Seems that there are quite a few duplicated company profiles like `aha` 😯 and `aha 2` 😲
Might be a good idea to remove the duplicates.

**Comment by RadValentin** [comment]
#### Update
- Ran into a blocker, turns out the plugin I wanted to use `remark-lint-are-links-valid` is flaky when checking links and returns a lot of false positives. The thing it uses to check links is not good at dealing with `https`  or redirects <https://github.com/stevenvachon/broken-link-checker/issues/58>
- Second option was to use `remark-lint-no-dead-urls` but this plugin is overly enthusiastic and will check all the links in all the files at once. Requests end up overlapping and we still get false positives... but when its ran on a single file it does a good job.
- I wrote a script that throttles the plugin to only run on 8 files at once. Worked ok.
- LinkedIn is paranoid and will return [code 999](https://stackoverflow.com/questions/27231113/999-error-code-on-head-request-to-linkedin) if it thinks you might be trying to scrape/crawl it. No way around that so far.

**TL;DR** It's now possible to check if links are broken. LinkedIn links will always fail the check.

@dougaitken Can you try to run the tests and let me know if it works for you as well? I added instructions in the contributing guide.

**Comment by dougaitken** [comment]
Hey @RadValentin been thinking about this recently. This would be really helpful, but it adds a higher barrier for entry. Someone can't simply add a profile and submit.
I'll leave it open for if this becomes an option in the future though

**Comment by nylen** [comment]
I think there is a lot of value in something like this, but it's probably worth writing a simple validation script ourselves rather than using third-party libraries which have (apparently) unexpected behavior.

See https://github.com/remoteintech/remote-jobs/pull/375 for an alternative approach based on converting Markdown to HTML and then inspecting the resulting structure.  Linting the Markdown files directly also seems valuable, but I think there are more and better tools for manipulating HTML, especially if we are trying to implement things unique to this repository.

> This would be really helpful, but it adds a higher barrier for entry. Someone can't simply add a profile and submit.

It seems like a good idea to me to have automated checks for broken links, especially when profiles are first submitted.  We could also set this up as a warning rather than an error.

**Comment by RadValentin** [comment]
> This would be really helpful, but it adds a higher barrier for entry. Someone can't simply add a profile and submit.

Agreed, [my plan](https://github.com/remoteintech/remote-jobs/issues/366#issuecomment-349129269) was for this test to be ran via a CI tool so the user doesn't need to set up anything, just check the CI logs.

This whole process could also be made optional or just kept as a tool that the repo maintainers run from time to time.

> I think there is a lot of value in something like this, but it's probably worth writing a simple validation script ourselves rather than using third-party libraries which have (apparently) unexpected behavior.

I want to offer an explanation as I think my previous comments might have given the wrong impression. Validating external URLs in general isn't a trivial process, it's hard to get accurate results because of the many factors in play (ISP, connection, hosts, etc.). 

This has nothing to do with [`remark`](https://github.com/gnab/remark) **parsing and linting markdown**.  Checking for broken links and linting are two separate concerns.

**Comment by dougaitken** [comment]
@RadValentin I'm going to pass on this for a bit, though I really appreciate you work. 

@nylen I'll bring in your side and we can discuss the next step for having a front end search

**Comment by RadValentin** [comment]
No prob

**Comment by RadValentin** [comment]
@nylen It would be awesome if in the future you could do a better job at communicating your intent ahead of time. If you plan to work on something then make an issue for it so other people know what's going on. Maybe open up the conversation. I would have liked to discuss linting for this repo and find a way for both of our solutions to work in tandem.

**Comment by nylen** [comment]
Sorry @RadValentin, I didn't see this PR until after I submitted my first one.

I think the main goal for now is to make this repo able to generate a front-end site; #377 is another step in this direction.

I think we can also achieve some nice wins for validation and automation along the way, but let's do it in a way that doesn't cause too much friction for maintainers or PR submitters (requiring to run tests themselves, etc).

I would like to collaborate on further work here as well.  @dougaitken has mentioned creating some issues for next steps, let's talk there about the best way to move forward.

**Comment by RadValentin** [comment]
Sounds good 👍 
I'll keep an eye open for future issues and see if there's any way I can help out.

**Comment by nylen** [comment]
Here are my thoughts on ways to follow up on this PR specifically.  Let me know if they sound reasonable.

## Link validation

I think we could check for broken links by sending `HEAD` requests (up to N in parallel) and special handling where needed (for LinkedIn, probably no validation).  [`request`](https://github.com/request/request) can be configured to handle redirects and HTTPS however we might want; would this approach be more fragile than I'm expecting?

This should probably also be a manual task because it's not deterministic (we don't want to fail people's PRs for unrelated changes or due to fragility), and it could take a while to go through every URL in the repository.

## Other misc stuff

I would think that fleshing out the `package.json` as you did here would be good, though I'm not sure what we would actually publish to npm from this repo, if anything.

Making some mention of the CI setup in the contributing guidelines would be good, though this could wait until we have a bot to leave comments on PRs.

**Comment by RadValentin** [comment]
**Link validation** seems right. One caveat are email links which should be ignored, shouldn't be a problem to filter them out. 

Another is with internal URLs, these can be checked without a request and this validation should happen on every PR (don't want to commit broken links to company profiles). This is what `remark-validate-links` was used for, we'll have to replicate this functionality. I'll make an issue for this.

Probably OK to leave `package.json` as is, you're right that those fields are only useful if publishing a package which doesn't make sense here.

Would be cool to add CI as soon as possible to help contributors, can you make an issue for this?

**Comment by dougaitken** [comment]
CI is now part of the PR process and does some checks @radvalentin, check out the latest for the wirecutter to see it in action

**Comment by nylen** [comment]
> Another is with internal URLs, these can be checked without a request and this validation should happen on every PR (don't want to commit broken links to company profiles).

This piece is now fully done, there were a bunch of PRs but now this code lives in `lib/index.js` with the other validation logic.

@RadValentin are you interested in looking at external link validation again?  I think it would be nice to have.

Here is how you can use the existing code to get a data structure for the repo's content, which may or may not be easier than parsing markdown:  https://github.com/remoteintech/remote-jobs/issues/88#issuecomment-399857201

**Comment by RadValentin** [comment]
@nylen I've moved on to other things and can't spare the time to look into this anymore. Feel free to tackle it if you want.
